### ISSUE TITLE
cannot get salt to replace a file

### ISSUE THREAD

**Issue by gzcwnk** [saltstack/salt#31938]
What am i doing wrong please?  for some reason the file does not get added to the server...
# my code is,

``` sls
rsyslog-logsetup:
  cmd.run:
    - name: 'mkdir -p /var/rsyslog'

rsyslog-backup:
  cmd.run:
    - name: 'mv /etc/rsyslog.conf /etc/orig-rsyslog.conf'

syslog-backup:
  cmd.run:
    - name: 'mv /etc/syslog.conf /etc/orig-syslog.conf'

#rhel6
{% if grains['oscodename'] == 'Santiago' %}

/etc/rsyslog.conf:
  file.managed:
    - source: salt://6-rsyslog2.conf
    - user: root
    - group: root
    - mode: 644
    - backup: minion

{% endif %}

#rhel7
{% if grains['oscodename'] == 'Maipo' %}
/etc/rsyslog.conf:
  file.managed:
    - source: salt://7-rsyslog2.conf
    - user: root
    - group: root
    - mode: 644
    - backup: minion

{% endif %}


#rhel5
{% if grains['oscodename'] == 'Tikanga' %}
syslog-restart:
  cmd.run:
    - name: 'service syslog restart'
{% endif %}

#rhel6
{% if grains['oscodename'] == 'Santiago' %}
rsyslog-restart:
  cmd.run:
    - name: 'service rsyslog restart'
{% endif %}
```
# 

run looks like this,
# 

```
-bash-4.1# salt -l debug -t 3600 vuwunicogsadbc1.ods.vuw.ac.nz state.sls rsyslog2
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: vuwunicorhsat03.vuw.ac.nz
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: vuwunicorhsat03.vuw.ac.nz
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'vuwunicorhsat03.vuw.ac.nz_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] LazyLoaded local_cache.get_load
[DEBUG   ] get_iter_returns for jid 20160317145908494197 sent to set(['vuwunicogsadbc1.ods.vuw.ac.nz']) will timeout at 15:59:08.518709
[DEBUG   ] jid 20160317145908494197 return from vuwunicogsadbc1.ods.vuw.ac.nz
[DEBUG   ] LazyLoaded highstate.output
[DEBUG   ] LazyLoaded nested.output
[DEBUG   ] LazyLoaded nested.output
[DEBUG   ] LazyLoaded nested.output
vuwunicogsadbc1.ods.vuw.ac.nz:
----------
          ID: rsyslog-logsetup
    Function: cmd.run
        Name: mkdir -p /var/rsyslog
      Result: True
     Comment: Command "mkdir -p /var/rsyslog" run
     Started: 14:59:08.813101
    Duration: 17.198 ms
     Changes:   
              ----------
              pid:
                  13118
              retcode:
                  0
              stderr:
              stdout:
----------
          ID: rsyslog-backup
    Function: cmd.run
        Name: mv /etc/rsyslog.conf /etc/orig-rsyslog.conf
      Result: False
     Comment: Command "mv /etc/rsyslog.conf /etc/orig-rsyslog.conf" run
     Started: 14:59:08.830640
    Duration: 16.528 ms
     Changes:   
              ----------
              pid:
                  13119
              retcode:
                  1
              stderr:
                  mv: cannot stat '/etc/rsyslog.conf': No such file or directory
              stdout:
----------
          ID: syslog-backup
    Function: cmd.run
        Name: mv /etc/syslog.conf /etc/orig-syslog.conf
      Result: False
     Comment: Command "mv /etc/syslog.conf /etc/orig-syslog.conf" run
     Started: 14:59:08.847457
    Duration: 16.562 ms
     Changes:   
              ----------
              pid:
                  13120
              retcode:
                  1
              stderr:
                  mv: cannot stat '/etc/syslog.conf': No such file or directory
              stdout:

Summary for vuwunicogsadbc1.ods.vuw.ac.nz
------------
Succeeded: 1 (changed=3)
Failed:    2
------------
Total states run:     3
[DEBUG   ] jid 20160317145908494197 found all minions set(['vuwunicogsadbc1.ods.vuw.ac.nz'])
ERROR: Minions returned with non-zero exit code
```
#

**Comment by gzcwnk** [comment]
[root@vuwunicogsadbc1 etc]# rpm -q salt-minion
salt-minion-2015.8.7-1.el7.noarch
[root@vuwunicogsadbc1 etc]#

====
-bash-4.1# rpm -q salt-master
salt-master-2015.8.7-1.el6.noarch
-bash-4.1#

**Comment by gzcwnk** [comment]
running salt-minion -l debug I can see my code comes up as blank?

==========
8><--------
[DEBUG   ] Rendered data from file: /var/cache/salt/minion/files/base/rsyslog2.sls:
rsyslog-logsetup:
  cmd.run:
    - name: 'mkdir -p /var/rsyslog'

rsyslog-backup:
  cmd.run:
    - name: 'mv /etc/rsyslog.conf /etc/orig-rsyslog.conf'

syslog-backup:
  cmd.run:
    - name: 'mv /etc/syslog.conf /etc/orig-syslog.conf'

#rhel6


#rhel7


#rhel5



#rhel5


#rhel6


#rhel7
8><----------
======

**Comment by gzcwnk** [comment]
======
-bash-4.1# salt -l debug -t 3600 vuwunicogsadbc1.ods.vuw.ac.nz grains.item os osrelease oscodename
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: vuwunicorhsat03.vuw.ac.nz
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: vuwunicorhsat03.vuw.ac.nz
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'vuwunicorhsat03.vuw.ac.nz_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] LazyLoaded local_cache.get_load
[DEBUG   ] get_iter_returns for jid 20160317153915717581 sent to set(['vuwunicogsadbc1.ods.vuw.ac.nz']) will timeout at 16:39:15.736055
[DEBUG   ] jid 20160317153915717581 return from vuwunicogsadbc1.ods.vuw.ac.nz
[DEBUG   ] LazyLoaded nested.output
vuwunicogsadbc1.ods.vuw.ac.nz:
    ----------
    os:
        RedHat
    oscodename:
        Red Hat Enterprise Linux Server 7.2 (Maipo)
    osrelease:
        7.2
[DEBUG   ] jid 20160317153915717581 found all minions set(['vuwunicogsadbc1.ods.vuw.ac.nz'])
-bash-4.1# 
============

**Comment by jfindlay** [comment]
@gzcwnk, thanks for reporting.  You can see the error coming from the shell in your output: `mv: cannot stat '/etc/rsyslog.conf': No such file or directory`.  Can you verify the presence of the file?  Also, the states you have enclosed in jinja will be blank if the platform you have run them on does not match the conditions.  You can test this without running it with `salt '*' state.show_sls rsyslog2`.

**Comment by gzcwnk** [comment]
Hi I mv the file out of the way otherwise the file.manage doesnt seem to swap in the new one.

**Comment by gzcwnk** [comment]
How do I specify  "maipo" only?  this used to work OK I think.

**Comment by gzcwnk** [comment]
It seems that,  "{% if grains['oscodename'] == 'Maipo' %}"  used to return Maipo which worked fine, now the returned string is longer and specifies 7.2 which is too specific.   What I need is the right syntax to only use Maipo otherwise for every RH release I will have to add a new search string which is silly.

**Comment by jfindlay** [comment]
@gzcwnk, you could try filtering on `osmajorrelease`: `{% if grains['osmajorrelease'] == 7 %}`.

**Comment by jfindlay** [comment]
@gzcwnk, it is possible that RedHat changed the information that `oscodename` is derived from.  You could also try,
```sls
{% if 'Maipo' in grains['oscodename'] %}
```

**Comment by gzcwnk** [comment]
I suspect RH have.

I am getting an error but cannot understand what it is saying,
===========
-bash-4.1# salt -l debug -t 3600 vuwunicogsadbc1.ods.vuw.ac.nz state.sls rsyslog2
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: vuwunicorhsat03.vuw.ac.nz
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: vuwunicorhsat03.vuw.ac.nz
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'vuwunicorhsat03.vuw.ac.nz_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] LazyLoaded local_cache.get_load
[DEBUG   ] get_iter_returns for jid 20160322085941846696 sent to set(['vuwunicogsadbc1.ods.vuw.ac.nz']) will timeout at 09:59:41.857220
[DEBUG   ] jid 20160322085941846696 return from vuwunicogsadbc1.ods.vuw.ac.nz
[DEBUG   ] LazyLoaded highstate.output
vuwunicogsadbc1.ods.vuw.ac.nz:
    Data failed to compile:
----------
    Rendering SLS 'base:rsyslog2' failed: Jinja syntax error: Unexpected end of template. Jinja was looking for the following tags: 'elif' or 'else' or 'endif'. The innermost block that needs to be closed is 'if'.; line 59

---
[...]
#rhel7
{% if grains['oscodename'] == 'Maipo' %}
rsyslog-restart:
  cmd.run:
    - name: 'systemctl restart rsyslog.service'
{% endif %}    <======================





[...]
---
[DEBUG   ] jid 20160322085941846696 found all minions set(['vuwunicogsadbc1.ods.vuw.ac.nz'])
ERROR: Minions returned with non-zero exit code
=========

My code is,
=========
rsyslog-logsetup:
  cmd.run:
    - name: 'mkdir -p /var/rsyslog'

rsyslog-backup:
  cmd.run:
    - name: 'mv /etc/rsyslog.conf /etc/orig-rsyslog.conf'

syslog-backup:
  cmd.run:
    - name: 'mv /etc/syslog.conf /etc/orig-syslog.conf'

#rhel6
{% if grains['oscodename'] == 'Santiago' %}

/etc/rsyslog.conf:
  file.managed:
    - source: salt://6-rsyslog2.conf
    - user: root
    - group: root
    - mode: 644
    - backup: minion

{% endif %}

#rhel7
#{% if grains['oscodename'] == 'Maipo' %}
{% if grains['os_family'] == 'RedHat' and grains['osmajorrelease'] == 7 %}
/etc/rsyslog.conf:
  file.managed:
    - source: salt://7-rsyslog2.conf
    - user: root
    - group: root
    - mode: 644
    - backup: minion

{% endif %}


#rhel5
{% if grains['oscodename'] == 'Tikanga' %}
syslog-restart:
  cmd.run:
    - name: 'service syslog restart'
{% endif %}

#rhel6
{% if grains['oscodename'] == 'Santiago' %}
rsyslog-restart:
  cmd.run:
    - name: 'service rsyslog restart'
{% endif %}

#rhel7
{% if grains['oscodename'] == 'Maipo' %}
rsyslog-restart:
  cmd.run:
    - name: 'systemctl restart rsyslog.service'
{% endif %}
=========

Is there anyway to make the debugging output better?

**Comment by gzcwnk** [comment]
neither of your suggestions work,

===========
rsyslog-logsetup:
  cmd.run:
    - name: 'mkdir -p /var/rsyslog'

rsyslog-backup:
  cmd.run:
    - name: 'mv /etc/rsyslog.conf /etc/orig-rsyslog.conf'

syslog-backup:
  cmd.run:
    - name: 'mv /etc/syslog.conf /etc/orig-syslog.conf'

#rhel7
{% if 'Maipo' in grains['oscodename'] %}
/etc/rsyslog.conf:
  file.managed:
    - source: salt://7-rsyslog2.conf
    - user:root
    - group:root
    - mode:644
    - backup:minion
{% endif %}
===========

===========
-bash-4.1# salt -l debug -t 3600 vuwunicogsadbc1.ods.vuw.ac.nz state.sls rsyslog2-test
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: vuwunicorhsat03.vuw.ac.nz
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] Configuration file path: /etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/master
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: vuwunicorhsat03.vuw.ac.nz
[DEBUG   ] Missing configuration file: /root/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/etc/salt/pki/master', 'vuwunicorhsat03.vuw.ac.nz_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] LazyLoaded local_cache.get_load
[DEBUG   ] get_iter_returns for jid 20160322092020525692 sent to set(['vuwunicogsadbc1.ods.vuw.ac.nz']) will timeout at 10:20:20.541345
[DEBUG   ] jid 20160322092020525692 return from vuwunicogsadbc1.ods.vuw.ac.nz
[DEBUG   ] LazyLoaded highstate.output
vuwunicogsadbc1.ods.vuw.ac.nz:
    Data failed to compile:
----------
    Too many functions declared in state 'file' in SLS u'rsyslog2-test'
[DEBUG   ] jid 20160322092020525692 found all minions set(['vuwunicogsadbc1.ods.vuw.ac.nz'])
ERROR: Minions returned with non-zero exit code
-bash-4.1# vi rsyslog2-test.sls
-bash-4.1# 
============

**Comment by gzcwnk** [comment]
Looking at 6.7 v 7.2 I can see the return differs,

-bash-4.1# salt vuwunicopatch06.ods.vuw.ac.nz grains.item os osrelease oscodename
vuwunicopatch06.ods.vuw.ac.nz:
    ----------
    os:
        RedHat
    oscodename:
        Santiago
    osrelease:
        6.7
-bash-4.1# salt vuwunicomysqlp2.ods.vuw.ac.nz grains.item os osrelease oscodename
vuwunicomysqlp2.ods.vuw.ac.nz:
    ----------
    os:
        RedHat
    oscodename:
        Red Hat Enterprise Linux Server 7.2 (Maipo)
    osrelease:
        7.2
-bash-4.1#

I reset my code to use "Red Hat Enterprise Linux Server 7.2 (Maipo)" and it works fine so its down to being unable to slice up the return and act on it.

:(

**Comment by jfindlay** [comment]
Your first error is that you have a jinja syntax error.  If you want to comment out jinja code, you need to use jinja comments `{#` `#}`.
```sls
rsyslog-logsetup:
  cmd.run:
    - name: 'mkdir -p /var/rsyslog'

rsyslog-backup:
  cmd.run:
    - name: 'mv /etc/rsyslog.conf /etc/orig-rsyslog.conf'

syslog-backup:
  cmd.run:
    - name: 'mv /etc/syslog.conf /etc/orig-syslog.conf'

#rhel6
{% if grains['oscodename'] == 'Santiago' %}

/etc/rsyslog.conf:
  file.managed:
    - source: salt://6-rsyslog2.conf
    - user: root
    - group: root
    - mode: 644
    - backup: minion

{% endif %}

#rhel7
{#{% if grains['oscodename'] == 'Maipo' %}#}
{% if 'Maipo' in grains['oscodename'] %}
/etc/rsyslog.conf:
  file.managed:
    - source: salt://7-rsyslog2.conf
    - user: root
    - group: root
    - mode: 644
    - backup: minion

{% endif %}


#rhel5
{% if grains['oscodename'] == 'Tikanga' %}
syslog-restart:
  cmd.run:
    - name: 'service syslog restart'
{% endif %}

#rhel6
{% if grains['oscodename'] == 'Santiago' %}
rsyslog-restart:
  cmd.run:
    - name: 'service rsyslog restart'
{% endif %}

#rhel7
{% if 'Maipo' in grains['oscodename'] %}
rsyslog-restart:
  cmd.run:
    - name: 'systemctl restart rsyslog.service'
{% endif %}
```

**Comment by gzcwnk** [comment]
thanks....no ideas on how I get to just "Maipo"  ?  the full thing "Red Hat Enterprise Linux Server 7.2 (Maipo) " works but when we go to 7.3 it will break again.

**Comment by jfindlay** [comment]
@gzcwnk, would you mind filing a new issue about the `oscodename` changing between I'm guessing 7.1 and 7.2?  It should be only `Maipo`.

**Comment by jfindlay** [comment]
Actually nevermind.  I think that that may already have been documented and fixed in #31365.
